File: app/src/main/java/com/shijc/wanandroidrx/ui/home/adapter/HomeAdapter.java
package com.shijc.wanandroidrx.ui.home.adapter;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import com.shijc.wanandroidrx.R;
import com.shijc.wanandroidrx.ui.home.bean.ArticleModel;
import com.shijc.wanandroidrx.ui.home.bean.BannerModel;
import com.shijc.wanandroidrx.utils.GlideImageLoader;
import com.shijc.wanandroidrx.utils.TimeUtils;
import com.youth.banner.Banner;
import com.youth.banner.listener.OnBannerListener;

import java.util.ArrayList;
import java.util.List;

/**
 * @author shijiacheng
 * @version V1.0
 * @Package com.shijc.wanandroidrx.ui.home.adapter
 * @Description:
 * @date 2019/4/3 上午 8:48
 */
public class HomeAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {




    private Context context;
    private List<ArticleModel> data;

    public HomeAdapter(Context context, List<ArticleModel> data) {
        this.context = context;
        this.data = data;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (viewType == BANNER){
            return new BannerViewHolder(LayoutInflater.from(context).inflate(R.layout.item_home_banner, parent, false));
        }else {
            return new RecyclerViewHolder(LayoutInflater.from(context).inflate(R.layout.item_rv_article, parent, false));
        }

    }

    @Override
    public int getItemCount() {
        return data.size();
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, final int position) {

        if (getItemViewType(position) == BANNER){
            if (holder instanceof BannerViewHolder) {
                //设置图片加载器
                ((BannerViewHolder) holder).banner.setImageLoader(new GlideImageLoader());

                if (banner.isEmpty()){
                    List<Integer> l = new ArrayList<Integer>();
                    l.add(R.drawable.bg1);
                    ((BannerViewHolder) holder).banner.setImages(l);
                }else{
                    //设置图片集合
                    List<String> list = new ArrayList<String>();
                    List<String> titles = new ArrayList<String>();
                    for (BannerModel.Data item : banner){
                        list.add(item.getImagePath());
                        titles.add(item.getTitle());
                    }
                    ((BannerViewHolder) holder).banner.setImages(list);
                    ((BannerViewHolder) holder).banner.setBannerTitles(titles);

                    ((BannerViewHolder) holder).banner.setOnBannerListener(new OnBannerListener() {
                        @Override
                        public void OnBannerClick(int position) {
                            if (listener!=null){
                                listener.onBannerClick(banner.get(position),position);
                            }
                        }
                    });
                }
                //banner设置方法全部调用完毕时最后调用
                ((BannerViewHolder) holder).banner.start();

            }
        }else {
            if (holder instanceof RecyclerViewHolder) {
                ((RecyclerViewHolder) holder).tvAuthor.setText(data.get(position-1).getAuthor());
                ((RecyclerViewHolder) holder).tvTime.setText(TimeUtils.long2String(data.get(position-1).getPublishTime(), TimeUtils.FORMAT_TYPE_1));
                ((RecyclerViewHolder) holder).tvContent.setText(data.get(position-1).getTitle());
                ((RecyclerViewHolder) holder).tvClassify.setText(data.get(position-1).getSuperChapterName());

                holder.itemView.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        if (listener!=null){
                            listener.onItemClick(data.get(position-1),position-1,v);
                        }
                    }
                });
            }
        }

    }


    class RecyclerViewHolder extends RecyclerView.ViewHolder {

        TextView tvAuthor;
        TextView tvTime;
        TextView tvContent;
        TextView tvClassify;

        public RecyclerViewHolder(View itemView) {
            super(itemView);
            tvAuthor = itemView.findViewById(R.id.tv_author);
            tvTime = itemView.findViewById(R.id.tv_time);
            tvContent = itemView.findViewById(R.id.tv_content);
            tvClassify = itemView.findViewById(R.id.tv_classify);
        }
    }


    class BannerViewHolder extends RecyclerView.ViewHolder {

        Banner banner;

        public BannerViewHolder(View itemView) {
            super(itemView);
            banner = itemView.findViewById(R.id.banner);
        }
    }







    private List<BannerModel.Data> banner = new ArrayList<BannerModel.Data>();

    public void setBanner(List<BannerModel.Data> banner){
        this.banner = banner;
        notifyDataSetChanged();
    }

    private static int BANNER = 0;
    private static int NORMAL = 1;


    @Override
    public int getItemViewType(int position) {

        if (position == 0){
            return BANNER;
        }else {
            return NORMAL;
        }

    }



    private ClickListener listener;

    public void setListener(ClickListener listener) {
        this.listener = listener;
    }

    public interface ClickListener{
        void onBannerClick(BannerModel.Data item ,int position);
        void onItemClick(ArticleModel item, int position, View view );
    }
}

File: app/src/main/java/com/shijc/wanandroidrx/ui/home/adapter/HomeAdapter.kt
package com.shijc.wanandroidrx.ui.home.adapter

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.shijc.wanandroidrx.R
import com.shijc.wanandroidrx.ui.home.bean.ArticleModel
import com.shijc.wanandroidrx.ui.home.bean.BannerModel
import com.shijc.wanandroidrx.utils.GlideImageLoader
import com.shijc.wanandroidrx.utils.TimeUtils
import com.youth.banner.Banner


/**
 * @Package com.shijc.wanandroidkotlin.ui.home.adapter
 * @Description:
 * @author shijiacheng
 * @date 2019/2/13 下午 4:52
 * @version V1.0
 */
class HomeAdapter(
    private val context: Context,
    private val data: List<ArticleModel>

) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private var banner: List<BannerModel.Data> = ArrayList<BannerModel.Data>()

    fun setBanner(banner: List<BannerModel.Data>){
        this.banner = banner
        notifyDataSetChanged()
    }

    private val BANNER = 0
    private val NORMAL = 1

    override fun getItemCount(): Int = data.size


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder = when (viewType) {
        BANNER -> {
            BannerViewHolder(
                LayoutInflater.from(context).inflate(R.layout.item_home_banner, parent, false)
            )
        }
        else -> {
            RecyclerViewHolder(
                LayoutInflater.from(context)
                    .inflate(R.layout.item_rv_article, parent, false)
            )
        }

    }

    /**
     * 利用kotlin中的when代替Java中的if/else
     * 返回不同的item类型
     */
    override fun getItemViewType(position: Int): Int = when (position) {
        0 -> {
            BANNER
        }
        else -> {
            NORMAL
        }
    }

    /**
     * 根据不同的item类型绑定不同的数据
     */
    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        when (getItemViewType(position)) {
            BANNER -> {
                if (holder is BannerViewHolder) {
                    //设置图片加载器
                    holder.banner.setImageLoader(GlideImageLoader())

                    if (banner.isEmpty()){
                        holder.banner.setImages(listOf(R.drawable.bg1))
                    }else{
                        //设置图片集合
                        val list = ArrayList<String>()
                        val titles = ArrayList<String>()
                        for (item in banner){
                            list.add(item.imagePath)
                            titles.add(item.title)
                        }
                        holder.banner.setImages(list)
                        holder.banner.setBannerTitles(titles)

                        holder.banner.setOnBannerListener {
                            if (listener!=null){
                                listener!!.onBannerClick(banner[it],it)
                            }
                        }
                    }
                    //banner设置方法全部调用完毕时最后调用
                    holder.banner.start()

                }
            }
            else -> {
                if (holder is RecyclerViewHolder) {
                    holder?.tvAuthor?.text = data[position-1].author
                    holder?.tvTime?.text = TimeUtils.long2String(data[position-1].publishTime, TimeUtils.FORMAT_TYPE_1)
                    holder?.tvContent?.text = data[position-1].title
                    holder?.tvClassify?.text = data[position-1].superChapterName

                    holder.itemView.setOnClickListener {
                        if (listener!=null){
                            listener!!.onItemClick(data[position-1],position-1,it)
                        }
                    }
                }
            }
        }
    }


    class BannerViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        var banner: Banner = view.findViewById(R.id.banner)
    }

    inner class RecyclerViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        var tvAuthor: TextView = view.findViewById(R.id.tv_author)
        var tvTime: TextView = view.findViewById(R.id.tv_time)
        var tvContent: TextView = view.findViewById(R.id.tv_content)
        var tvClassify: TextView = view.findViewById(R.id.tv_classify)
    }

    var listener:ClickListener? = null

    interface ClickListener{
        fun onBannerClick(item: BannerModel.Data,position: Int);
        fun onItemClick(item: ArticleModel, position:Int, view: View)
    }
}